File: quizapp/app/src/main/java/uwr/dbozek/quizapp/MainActivity.java
package uwr.dbozek.quizapp;

import androidx.appcompat.app.AppCompatActivity;
import androidx.cardview.widget.CardView;
import androidx.constraintlayout.widget.ConstraintLayout;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;
import android.widget.Toast;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class MainActivity extends AppCompatActivity implements RadioGroup.OnCheckedChangeListener {

    private List<Question> allQuestions = Arrays.asList(
            new Question("What is the SI unit of force?", Arrays.asList("A) Newton (N)", "B) Joule (J)", "C) Watt (W)", "D) Pascal (Pa)"), 0),
            new Question("What is the formula for acceleration?", Arrays.asList("A) v = u + at", "B) F = ma", "C) E = mc^2", "D) F = G(m1*m2/r^2)"), 0),
            new Question("Which of the following is a vector quantity?", Arrays.asList("A) Speed", "B) Distance", "C) Velocity", "D) Mass"), 2),
            new Question("What is the conservation of energy?", Arrays.asList("A) Energy cannot be created nor destroyed, only transformed", "B) Energy can be created from nothing", "C) Energy can be destroyed", "D) None of the above"), 0),
            new Question("What is the speed of light in a vacuum?", Arrays.asList("A) 299,792,458 meters per second", "B) 300,000 meters per second", "C) 3,000,000 meters per second", "D) 30,000 meters per second"), 0),
            new Question("Which sorting algorithm has the best average and worst-case time complexity?", Arrays.asList("A) Quick Sort", "B) Bubble Sort", "C) Merge Sort", "D) Insertion Sort"), 0),
            new Question("What is the time complexity of binary search?", Arrays.asList("A) O(n)", "B) O(log n)", "C) O(n^2)", "D) O(1)"), 1),
            new Question("What is the purpose of dynamic programming?", Arrays.asList("A) To efficiently solve problems with overlapping subproblems", "B) To efficiently solve problems with non-overlapping subproblems", "C) To solve problems with constant time complexity", "D) None of the above"), 0),
            new Question("What is the Big O notation for linear search?", Arrays.asList("A) O(n)", "B) O(log n)", "C) O(n^2)", "D) O(1)"), 0),
            new Question("Which data structure uses LIFO (Last In, First Out) principle?", Arrays.asList("A) Stack", "B) Queue", "C) Linked List", "D) Tree"), 0),
            new Question("What is the formula for momentum?", Arrays.asList("A) p = m * v", "B) F = ma", "C) E = mc^2", "D) F = G(m1*m2/r^2)"), 0),
            new Question("Which law of thermodynamics states that energy cannot be created nor destroyed, only transformed?", Arrays.asList("A) First Law", "B) Second Law", "C) Third Law", "D) Zeroth Law"), 0),
            new Question("What is the unit of electric charge?", Arrays.asList("A) Coulomb (C)", "B) Ampere (A)", "C) Volt (V)", "D) Ohm (Ω)"), 0),
            new Question("Which algorithm is used for finding the shortest path in a weighted graph?", Arrays.asList("A) Dijkstra's Algorithm", "B) Depth-First Search (DFS)", "C) Breadth-First Search (BFS)", "D) Prim's Algorithm"), 0),
            new Question("What is the SI unit of power?", Arrays.asList("A) Watt (W)", "B) Joule (J)", "C) Newton (N)", "D) Volt (V)"), 0),
            new Question("What is the formula for work done?", Arrays.asList("A) W = Fd cos(θ)", "B) W = mgh", "C) W = F/a", "D) W = QV"), 0),
            new Question("Which law of motion states that an object at rest tends to stay at rest, and an object in motion tends to stay in motion?", Arrays.asList("A) Newton's First Law", "B) Newton's Second Law", "C) Newton's Third Law", "D) Einstein's Theory of Relativity"), 0),
            new Question("What is the formula for gravitational potential energy?", Arrays.asList("A) U = mgh", "B) U = 1/2mv^2", "C) U = -G(m1*m2/r)", "D) U = kx^2"), 0),
            new Question("What is the value of gravitational acceleration on Earth?", Arrays.asList("A) 9.81 m/s²", "B) 3.00 x 10^8 m/s", "C) 6.67 x 10^-11 Nm^2/kg^2", "D) 1.00 atm"), 0),
            new Question("Which algorithm is used for finding the minimum spanning tree in a weighted graph?", Arrays.asList("A) Prim's Algorithm", "B) Dijkstra's Algorithm", "C) Bellman-Ford Algorithm", "D) Kruskal's Algorithm"), 0),
            new Question("What is the unit of resistance?", Arrays.asList("A) Ohm (Ω)", "B) Ampere (A)", "C) Volt (V)", "D) Watt (W)"), 0),
            new Question("Which algorithm is used for sorting a linked list?", Arrays.asList("A) Merge Sort", "B) Bubble Sort", "C) Quick Sort", "D) Selection Sort"), 0),
            new Question("What is the formula for centripetal force?", Arrays.asList("A) F = mv²/r", "B) F = ma", "C) F = G(m1*m2/r²)", "D) F = kx"), 0)

    );

    private TextView questionNumber, questionText, scoreText, finalScoreText;
    private ProgressBar progressBar;
    private RadioGroup answerRadioGroup;
    private Button nextButton, restartButton;
    private ConstraintLayout quizLayout;
    private LinearLayout congratsLayout;

    private CardView answerCardView;

    private List<Question> selectedQuestions;
    private int currentQuestionIndex = 0;
    private int score = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Inicjalizacja widoków
        questionNumber = findViewById(R.id.questionNumber);
        progressBar = findViewById(R.id.progressBar);
        questionText = findViewById(R.id.questionText);
        answerRadioGroup = findViewById(R.id.answerRadioGroup);
        nextButton = findViewById(R.id.nextButton);
        answerCardView = findViewById(R.id.answerCardView);
//        scoreText = findViewById(R.id.scoreText);

        congratsLayout = findViewById(R.id.congratsLayout);
        quizLayout = findViewById(R.id.quizLayout);

        finalScoreText = findViewById(R.id.finalScoreText);
        restartButton = findViewById(R.id.restartButton);

        // Losowanie 10 pytań z puli
        selectedQuestions = new ArrayList<>(allQuestions);
        Collections.shuffle(selectedQuestions);
        selectedQuestions = selectedQuestions.subList(0, 10);

        showQuestion(selectedQuestions.get(currentQuestionIndex));

        nextButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                checkAnswer();
            }
        });

        restartButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                restartQuiz();
            }
        });

        answerRadioGroup.setOnCheckedChangeListener(this);
    }

    private void showQuestion(Question question) {
        questionNumber.setText("Question " + (currentQuestionIndex + 1) + "/10");
        progressBar.setProgress(currentQuestionIndex + 1);
        questionText.setText(question.getQuestionText());

        answerRadioGroup.removeAllViews();
        for (String answer : question.getAnswers()) {
            RadioButton radioButton = new RadioButton(this);
            radioButton.setText(answer);
            answerRadioGroup.addView(radioButton);
        }

        // Blokowanie przycisku "Next" na początku każdego pytania
        nextButton.setEnabled(false);
    }

    private void checkAnswer() {
        int selectedAnswerId = answerRadioGroup.getCheckedRadioButtonId();

        if (selectedAnswerId == -1) {
            Toast.makeText(this, "Please select an answer", Toast.LENGTH_SHORT).show();
            return; // Return early if no answer is selected
        }

        int selectedAnswerIndex = answerRadioGroup.indexOfChild(
                findViewById(selectedAnswerId));

        if (selectedAnswerIndex == selectedQuestions.get(currentQuestionIndex).getCorrectAnswerIndex()) {
            score += 10;
        }

        currentQuestionIndex++;

        if (currentQuestionIndex < selectedQuestions.size()) {
            showQuestion(selectedQuestions.get(currentQuestionIndex));
            // Blokowanie przycisku "Next" na początku każdego pytania
            nextButton.setEnabled(false);
        } else {
            showFinalScore();
        }
    }

    private void showFinalScore() {
//        quizLayout.setVisibility(View.GONE);

        questionNumber.setVisibility(View.GONE);
        progressBar.setVisibility(View.GONE);
        questionText.setVisibility(View.GONE);
        answerRadioGroup.setVisibility(View.GONE);
        nextButton.setVisibility(View.GONE);
        answerCardView.setVisibility(View.GONE);

        congratsLayout.setVisibility(View.VISIBLE);

        finalScoreText.setText("Your Score: " + score);
    }

    private void restartQuiz() {
        currentQuestionIndex = 0;
        score = 0;
        progressBar.setProgress(0);
//        quizLayout.setVisibility(View.VISIBLE);

        questionNumber.setVisibility(View.VISIBLE);
        progressBar.setVisibility(View.VISIBLE);
        questionText.setVisibility(View.VISIBLE);
        answerRadioGroup.setVisibility(View.VISIBLE);
        nextButton.setVisibility(View.VISIBLE);
        answerCardView.setVisibility(View.VISIBLE);

        congratsLayout.setVisibility(View.GONE);
        selectedQuestions = new ArrayList<>(allQuestions);
        Collections.shuffle(selectedQuestions);
        selectedQuestions = selectedQuestions.subList(0, 10);
        showQuestion(selectedQuestions.get(currentQuestionIndex));
    }

    @Override
    public void onCheckedChanged(RadioGroup group, int checkedId) {
        nextButton.setEnabled(true);
    }
}

File: QuizKotlin/app/src/main/java/uwr/dbozek/quizkt/MainActivity.kt
package uwr.dbozek.quizkt

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.RadioButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.input.TextFieldValue
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import uwr.dbozek.quizkt.ui.theme.QuizKtTheme
import java.util.Arrays
import java.util.Collections

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            MyApp()
        }
    }
}

data class Question(
    val questionText: String,
    val answers: List<String>,
    val correctAnswerIndex: Int
    )

@Composable
fun MyCard(text: String) {
    Card(
        modifier = Modifier
            .height(100.dp)
            .fillMaxWidth()

    ) {
        Text(text = text, modifier = Modifier.padding(16.dp), textAlign = TextAlign.Center, fontSize = 16.sp)
    }
}

@Composable
fun MyRadioButton(text: String){
    Row (modifier = Modifier
        .padding(vertical = 8.dp)
        .fillMaxWidth(), verticalAlignment = Alignment.CenterVertically){
        RadioButton(selected = true, onClick = { }, modifier = Modifier.padding(end = 8.dp))
        Text(text = text, fontSize = 14.sp)
    }
}




@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun MyApp(){

    var questionIndex by remember { mutableStateOf(0) }
    var selectedAnswerIndex by remember { mutableStateOf(-1) }
    var points by remember { mutableStateOf(0) }
    var showEndScreen by remember { mutableStateOf(false)}


    val allquestions: List<Question> = listOf(
        Question("What is the SI unit of force?", listOf("Newton (N)","Joule (J)","Watt (W)","Pascal (Pa)"),0),
        Question("What is the formula for acceleration?", listOf("v = u + at", "F = ma", "E = mc^2", "F = G(m1*m2/r^2)"), 0),
        Question("Which of the following is a vector quantity?", listOf("Speed", "Distance", "Velocity", "Mass"), 2),
        Question("What is the conservation of energy?", listOf("Energy cannot be created nor destroyed, only transformed", "Energy can be created from nothing", "Energy can be destroyed", "None of the above"), 0),
        Question("What is the Big O notation for linear search?", listOf("O(n)", "O(log n)", "O(n^2)", "O(1)"), 0),
        Question("Which data structure uses LIFO (Last In, First Out) principle?", listOf("Stack", "Queue", "Linked List", "Tree"), 0),
        Question("What is the formula for work done?", listOf("W = Fd cos(θ)", "W = mgh", "W = F/a", "W = QV"), 0),
        Question("What is the unit of electrical resistance?", listOf("Ohm", "Ampere", "Volt", "Watt"), 0),
        Question("What is the formula for Ohm's Law?", listOf("V = IR", "P = IV", "V = I^2R", "P = V^2/R"), 0),
        Question("Which sorting algorithm has the best average time complexity?", listOf("Bubble Sort","Merge Sort","Quick Sort","Selection Sort"), 1),
        Question("What does HTTP stand for in web development?", listOf("Hypertext Transfer Protocol","Hypertext Transmission Protocol","Hypertext Terminal Process","Hypertext Transfer Process"), 0),
        Question("What is the function of a firewall?", listOf("To enhance computer processing speed","To create network connections","To protect a network from unauthorized access","To debug programming code"), 2),
        Question("What does IDE stand for in software development?", listOf("Integrated Development Environment","Integrated Design Environment","Integrated Debugging Environment","Interchangeable Development Environment"), 0),
        Question("What is the purpose of inheritance in object-oriented programming?", listOf("To restrict access to certain class members","To allow a new class to inherit properties and behavior from an existing class","To create multiple instances of a class","To define the structure of a class"), 1)

    )

    var selectedQuestions: List<Question> = allquestions
    Collections.shuffle(selectedQuestions)
    selectedQuestions = selectedQuestions.subList(0,10)


    val currentQuestion = selectedQuestions[questionIndex]

    val onAnswerSelected: (Int) ->Unit = { answerIndex -> selectedAnswerIndex = answerIndex}
    val progress = ((questionIndex + 1) / selectedQuestions.size.toFloat()).coerceIn(0f,1f)



    if(showEndScreen){
        EndScreen(points = points) {
            showEndScreen = false
            points = 0
        }
    }else{
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(34.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ){
            Text(text = "Question ${questionIndex+1} of 10",
                fontSize = 30.sp,
                color = Color.Black,
            )

            LinearProgressIndicator(
                progress = progress,
                modifier = Modifier.fillMaxWidth()
            )
            MyCard(text = currentQuestion.questionText)

            Card(
                modifier = Modifier.fillMaxWidth()

            ) {
                Column(
                    modifier = Modifier.padding(16.dp),
                    verticalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    currentQuestion.answers.forEachIndexed { index, answer ->

                        Card (
                            modifier = Modifier
                                .padding(12.dp)
                                .padding(vertical = 2.dp)
                                .fillMaxWidth(),
                            shape = RoundedCornerShape(8.dp),
                            colors = CardDefaults.cardColors(
                                containerColor = if (index == selectedAnswerIndex) Color(0xFFefd6ff) else Color(0xFFdbd6ff)
                            ),
                            onClick = {

                                onAnswerSelected(index)
                            }

                        ){
                            //MyRadioButton(text = answer)
                            Row (modifier = Modifier
                                .padding(vertical = 8.dp)
                                .fillMaxWidth(), verticalAlignment = Alignment.CenterVertically){
                                RadioButton(selected = if (index == selectedAnswerIndex){true}else{false}, onClick = { onAnswerSelected(index) }, modifier = Modifier.padding(end = 8.dp))
                                Text(text = answer, fontSize = 14.sp)
                            }


                        }

                    }
                }
            }
            Spacer(modifier = Modifier.weight(1f))
            Button(

                onClick = {
                    if(selectedAnswerIndex != -1) {
                        if(currentQuestion.correctAnswerIndex == selectedAnswerIndex){
                            points = points + 10
                        }

                        if(questionIndex < selectedQuestions.size - 1){
                            questionIndex++
                            selectedAnswerIndex = -1}
                        else{
                            showEndScreen = true
                            questionIndex = 0


                            selectedAnswerIndex = -1
                        }}
                },
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(vertical = 18.dp),
                shape = MaterialTheme.shapes.medium
            ) {
                Text(text = "Next",
                    fontSize = 18.sp)
            }
        }
    }
}




@Composable
fun EndScreen(points: Int, onRestart: () -> Unit) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(16.dp)
            .background(Color.White),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
    ){
        Text(
            text = "End of quiz!",
            fontSize = 24.sp,
            color = Color.Black
        )
        Text(
            text = "Points: $points",
            fontSize = 18.sp,
            color = Color.Black
        )
        Button(
            onClick = { onRestart() }    ,
            modifier = Modifier
                .fillMaxWidth()
                .padding(vertical = 18.dp),
            shape = MaterialTheme.shapes.medium) {
            Text(text = "Restart")
            
        }
    }
}



@Preview(showBackground = true, showSystemUi = true)
@Composable
fun Preview() {
   //MyApp()

}